### PR TITLE
Bootstrap 4, piece by WIP piece

### LINKED ISSUE #1632: Move to bootstrap 4 and bootstrap_form 4
Thanks for creating an issue! Please fill out this form so we can be sure to have all the information we need, and to minimize back and forth.

* **What are we trying to do?**

We're a major version behind on bootstrap, which wouldn't be a terribly huge deal except our form builder is also moving on, and I want to make sure we're taking advantage of what it offers us in terms of security. So we should move to 

* **What feature or behavior is this required for?**

Long term health of the project, rails security, the ability to use form_with in code.

* **How could we solve this issue? (Not knowing is okay!)**

At a minimum, `gem 'bootstrap'` and `bundle update bootstrap_form`, then work to make all the tests pass again and make sure the views don't get wild.

* **Anything else?**

This is going to be a lot of work, as our CSS is probably the least touched part of this project, but it is absolutely worth doing so we can keep our form building up to snuff.

### PR DESCRIPTION
**Pull request by colinxfleming**
here's a master PR for our bootstrap4 upgrade.

for now, let's approve the fix to the navbar and budget bar, and then other components/sections will be their own PRs.

Here's navbar:

in sandbox:

image

in bootstrap 4 (differences are that the active dropdown is no longer highlighted, and the navbar fills the width instead of the container):

image

fixes #1632 ... eventually

### CODE REVIEW

File: app/assets/javascripts/tooltips.coffee [hunk 1]
```diff
@@ -1,12 +1,7 @@
 # Pertaining to tooltips
 $(document).on 'turbolinks:load', ->
   # For regular spans
-  $('.daria-tooltip').each ->
-    $(@).tooltip({
-      html: true,
-      placement: 'bottom',
-      title: $(@).data('tooltip-text')
-    })
+  $('.daria-tooltip').tooltip()
```
Comment (colinxfleming) [inline app/assets/javascripts/tooltips.coffee]: setting this via data attributes is a little closer to the metal, so we do that there.

File: app/assets/stylesheets/application.css.scss [hunk 2]
```diff
@@ -5,7 +5,6 @@
  *= require_tree .
```
Comment (colinxfleming) [inline app/assets/stylesheets/application.css.scss]: I erred on the side of hacking things into the css here, on the assumption that we would eventually be reorganizing this and moving it to webpacker, but I AM leaving a mess here. So we could also clean it up a little bit as part of this process, start sticking global styles in their own sheet, etc. But it's also not strictly necessary for BS4, so I will leave this call up to the reviewer - hacky CSS okay, or should we start cleaning things up while we're wrenching around in here?

File: app/assets/stylesheets/application.css.scss [hunk 3]
```diff
@@ -199,17 +203,17 @@ h5 {
 // using Bootstrap
 // generated by the rails_layout gem
 .authform {
-  @extend .well;
-  @extend .well-lg;
-  @extend .center-block;
+  @extend .card-header;
+  // @extend .well-lg;
```
Comment (colinxfleming) [inline app/assets/stylesheets/application.css.scss]: they took out well and replaced it with cards.

File: app/helpers/navbar_helper.rb [hunk 4]
```diff
@@ -3,16 +3,16 @@ def cm_resources_link
...
-      link_to t('navigation.cm_resources.label'), url, target: '_blank'
+      link_to t('navigation.cm_resources.label'), url, target: '_blank', class: 'nav-link'
     end if url.present?
   end
 
   def spanish_or_english_link
     content_tag :li do
       if I18n.locale == I18n.default_locale
-        link_to "Español", { :locale=>'es' }
+        link_to "Español", { locale: 'es' }, class: 'nav-link'
```
Comment (colinxfleming) [inline app/helpers/navbar_helper.rb]: `nav-link` is what applies standard navbar styling now.

File: app/views/layouts/_navigation_links.html.erb [hunk 5]
```diff
@@ -1,39 +1,38 @@
...
-</ul>
+<% if current_user.admin? || current_user.data_volunteer? %>
+  <div class="dropdown">
+    <button href="#" data-toggle="dropdown" class="btn dropdown-toggle navbar-text" id="admin-dropdown" aria-haspopup="true" aria-expanded="false">
+      <%= t('navigation.admin_tools.label') %> 
+    </button>
+
+    <div class="dropdown-menu" aria-labelledby="admin-dropdown">
+      <% if current_user.admin? %>
+        <%= link_to t('navigation.admin_tools.user_management'), users_path, class: 'dropdown-item' %>
```
Comment (colinxfleming) [inline app/views/layouts/_navigation_links.html.erb]: lots of `li`s get replaced by `a`s in bs4.

File: app/assets/stylesheets/events.scss [hunk 6]
```diff
@@ -16,3 +16,9 @@
   height: 300px;
   overflow: auto;
 }
+
+// Styling for Kaminari pagination
+span.page, span.next, span.last {
```
Comment (colinxfleming) [inline app/assets/stylesheets/events.scss]: something really tries to smash these together, so we expand them a little bit with margin.

File: app/assets/stylesheets/application.css.scss (hunk 2 above)
Comment (lomky) [reply app/assets/stylesheets/application.css.scss]: This is a big enough chunk of work already that I'm pretty on board with Doing It Later, so long as we aren't making more work for us in the process.

File: Gemfile [hunk 7]
```diff
@@ -14,7 +14,8 @@ gem 'bootsnap', '>= 1.1.0', require: false
 # Asset pipeline
```
Comment (lomky) [inline Gemfile]: File :+1:

File: Gemfile.lock [hunk 8]
```diff
@@ -56,14 +56,15 @@ GEM
     afm (0.2.2)
```
Comment (lomky) [inline Gemfile.lock]: File +1

File: app/assets/javascripts/application.js [hunk 9]
```diff
@@ -12,6 +12,7 @@
 // about supported directives.
```
Comment (lomky) [inline app/assets/javascripts/application.js]: File :+1:

File: app/assets/javascripts/tooltips.coffee [hunk 10]
```diff
@@ -1,12 +1,7 @@
 # Pertaining to tooltips
```
Comment (lomky) [inline app/assets/javascripts/tooltips.coffee]: File :+1:

File: app/assets/stylesheets/events.scss [hunk 11]
```diff
@@ -16,3 +16,9 @@
   height: 300px;
```
Comment (lomky) [inline app/assets/stylesheets/events.scss]: File :+1:

File: app/helpers/budget_bar_helper.rb [hunk 12]
```diff
@@ -2,7 +2,7 @@
 module BudgetBarHelper
```
Comment (lomky) [inline app/helpers/budget_bar_helper.rb]: File :+1:

File: app/helpers/lines_helper.rb [hunk 13]
```diff
@@ -8,7 +8,10 @@ def lines
   end
```
Comment (lomky) [inline app/helpers/lines_helper.rb]: File :+1:

File: app/helpers/navbar_helper.rb [hunk 14]
```diff
@@ -3,16 +3,16 @@ def cm_resources_link
     url = Config.find_or_create_by(config_key: 'resources_url').options.first
```
Comment (lomky) [inline app/helpers/navbar_helper.rb]: File :+1:

File: app/helpers/tooltips_helper.rb [hunk 15]
```diff
@@ -2,8 +2,10 @@
 module TooltipsHelper
```
Comment (lomky) [inline app/helpers/tooltips_helper.rb]: File :+1:

File: app/views/dashboards/_budget_bar.html.erb [hunk 16]
```diff
@@ -1,40 +1,30 @@
-<div class="col-sm-12" id='budget-bar-content'>
```
Comment (lomky) [inline app/views/dashboards/_budget_bar.html.erb]: File :+1:

File: app/views/devise/sessions/new.html.erb [hunk 17]
```diff
@@ -1,42 +1,42 @@
...
-    <% end %>
-
-    <%= render "devise/shared/links" %>
-  </div>
-  <div class="navbar">
-    <ul class="nav navbar-nav navbar-left">
-      <%= spanish_or_english_link %>
-    </ul>
+<div class="row">
+  <div class="col-sm-6">
```
Comment (lomky) [inline app/views/devise/sessions/new.html.erb]: File :+1:

File: app/views/events/_events.html.erb [hunk 18]
```diff
@@ -1,13 +1,15 @@
-<div class='well well-lg' id='event-data'>
```
Comment (lomky) [inline app/views/events/_events.html.erb]: File :+1:

File: app/views/layouts/_navigation.html.erb [hunk 19]
```diff
@@ -1,16 +1,13 @@
-<nav class="navbar navbar-inverse navbar-fixed-top">
```
Comment (lomky) [inline app/views/layouts/_navigation.html.erb]: File :+1:

File: app/views/layouts/_navigation_links.html.erb [hunk 20]
```diff
@@ -1,39 +1,38 @@
-<ul class="nav navbar-nav navbar-left">
```
Comment (lomky) [inline app/views/layouts/_navigation_links.html.erb]: File :+1:

File: app/views/layouts/application.html.erb [hunk 21]
```diff
@@ -1,7 +1,7 @@
 <!DOCTYPE html>
```
Comment (lomky) [inline app/views/layouts/application.html.erb]: File :+1:

**Comment by colinxfleming** [comment]
preponderance of the work is done here! still remaining:
* data entry view
* pass thru some of the other views to make sure I didn't miss a spot
* pledge flow
* fulfillment partial
* main dashboard
* spacer classes, proper use of hidden classes, etc
* final pass with other developer eyes

**Comment by colinxfleming** [comment]
I think this is all done except for spacer classes and test failures. Working on test failures this weekend and hopefully we can put a fork in this soon.

**Comment by colinxfleming** [comment]
This is shippable, but it touches enough that we need a final review/spot check. Will plan on doing the final review myself shortly, and getting another second set of eyes.

**Comment by colinxfleming** [comment]
extremely regrettably, I think this needs a rebase.

File: app/assets/stylesheets/application.css.scss (hunk 2 above)
Comment (colinxfleming) [reply app/assets/stylesheets/application.css.scss]: popped this into #1854

**Pull request merged**